src/spatialmath/quaternion.rs

```diff
@@ -0,0 +1,371 @@
+        let quat = Quaternion::new(1.0, 0.0, 0.5, 1.0);
+        let expected_imag = Vector3::new(0.0, 0.5, 1.0);
+        assert_eq!(quat.imag(), expected_imag);
+    }
+
+    #[test]
+    fn set_imag_from_vector_works() {
+        let imag = Vector3::new(0.0, 0.5, 1.0);
+        let expected_quat = Quaternion::new(1.0, 0.0, 0.5, 1.0);
+        let quat = Quaternion::new_from_vector(1.0, imag);
+        assert_eq!(quat, expected_quat);
+    }
+
+    #[test]
+    fn quaternion_normalizes_successfully() {
+        let mut quat = Quaternion::new(
+            0.0,
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * -0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5
+        );
+        let expected_quat = Quaternion::new(
+            0.0,
+            (1.0_f64 / 3.0_f64).sqrt(), 
+            (1.0_f64 / 3.0_f64).sqrt() * -1.0, 
+            (1.0_f64 / 3.0_f64).sqrt()
+        );
+        quat.normalize();
+        assert!(approx_eq!(Quaternion, quat, expected_quat));
+    }
+
+    #[test]
+    fn is_normalized_is_correct() {
+        let not_normalized = Quaternion::new(
+            0.3,
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * -0.5, 
+            (1.0_f64 / 3.0_f64).sqrt() * 0.5
+        );
+        assert!(!not_normalized.is_normalized());
+        let normalized = Quaternion::new(
+            0.0,
+            (1.0_f64 / 3.0_f64).sqrt(), 
+            (1.0_f64 / 3.0_f64).sqrt() * -1.0, 
+            (1.0_f64 / 3.0_f64).sqrt()
+        );
+        assert!(normalized.is_normalized());
+    }
+
+    #[test]
+    fn add_subtract_works() {
+        let quat1 = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        let quat2 = Quaternion::new(0.2, 0.3, 0.4, 0.5);
+        let expected_add = Quaternion::new(0.3, 0.5, 0.7, 0.9);
+        let expected_sub = Quaternion::new(-0.1, -0.1, -0.1, -0.1);
+        assert!(approx_eq!(Quaternion, quat1 + quat2, expected_add));
+        assert!(approx_eq!(Quaternion, quat1 - quat2, expected_sub));
+    }
+
+    #[test]
+    fn multiply_works() {
+        let quat1 = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        let quat2 = Quaternion::new(0.2, 0.3, 0.4, 0.5);
+        let expected_mul = Quaternion::new(-0.36, 0.06, 0.12, 0.12);
+        let expected_rev_mul = Quaternion::new(-0.36, 0.08, 0.08, 0.14);
+        assert!(approx_eq!(Quaternion, quat1 * quat2, expected_mul));
+        assert!(approx_eq!(Quaternion, quat2 * quat1, expected_rev_mul));
+    }
+
+    #[test]
+    fn scale_works() {
+        let mut quat = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        quat.scale(2.0);
+        let expected_quat = Quaternion::new(0.2, 0.4, 0.6, 0.8);
+        assert!(approx_eq!(Quaternion, quat, expected_quat));
+    }
+
+    #[test]
+    fn get_scaled_returns_scaled_quaternion() {
+        let quat = Quaternion::new(0.1, 0.2, 0.3, 0.4);
+        let quat_orig = Quaternion::new(0.1, 0.2, 0.3, 0.4);
```

Comment:
Mild preference for `let quat_orig = quat.clone()` here. By defining it twice, it feels like we don't have a single point of truth on this thing.